In [106]:
# Imports and pip installations (if needed)
import pandas as pd
from sklearn.preprocessing import StandardScaler


# Part 1: Load the dataset

In [107]:
# Load the given datasets
ckd_categorical = pd.read_csv(".github/chronic_kidney_disease_categorical.csv")
ckd_numerical = pd.read_csv(".github/chronic_kidney_disease_numerical.csv")

# Print the data
print(ckd_categorical)
print(ckd_numerical)

     unique_id   al   su       rbc        pc         pcc          ba  htn  \
0       224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes   
1       992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   
2       308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   
3       450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   
4       881763  0.0  0.0    normal    normal  notpresent  notpresent   no   
..         ...  ...  ...       ...       ...         ...         ...  ...   
425     769512  0.0  0.0    normal    normal  notpresent  notpresent   no   
426     872034  2.0  1.0  abnormal  abnormal  notpresent     present  yes   
427     643007  NaN  NaN       NaN    normal  notpresent  notpresent  yes   
428     840392  NaN  NaN       NaN       NaN  notpresent  notpresent  yes   
429     856244  0.0  0.0       NaN    normal  notpresent  notpresent  yes   

      dm  cad appet   pe  ane  Target  
0    yes   no  good  yes   no     c

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer:

The Chronic Kidney Disease dataset is used to predict the presence of chronic kidney disease based on various features. Here's a breakdown of the dataset:

Features and Labels

Features: The dataset includes 24 features that describe different attributes related to health and kidney function.

Label: The target variable is "class," indicating whether a patient has chronic kidney disease (ckd) or not (notckd).

Types of Features

The dataset contains both categorical and numerical features:

Categorical Features:
Specific Gravity (sg): Levels like 1.005, 1.010, etc.
Albumin (al): Levels from 0 to 5.
Sugar (su): Levels from 0 to 5.

Numerical Features:
Age: Age in years.
Blood Pressure (bp): Measured in mm/Hg.
Blood Glucose Random (bgr): Measured in mg/dl.

Summary

The dataset is multivariate and used for classification tasks. It includes a mix of categorical and numerical data, providing a comprehensive overview of factors that may influence kidney health.

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [108]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
duplicates = ckd_numerical.duplicated(subset=['unique_id'])

# Count the total number of duplicate rows
num_duplicates = duplicates.sum()

# Print the total number of duplicate rows
if num_duplicates > 0:
    print(f"Total number of Numericsl duplicate rows: {num_duplicates}")
    
    # Drop these duplicate rows
    ckd_numerical = ckd_numerical.drop_duplicates(subset=['unique_id'])
    print("Numerical Duplicate rows removed.")
else:
    print("No Numerical duplicate rows found.")



# Repeat the same for categorical dataset. Print the duplicate rows and drop them
duplicates1 = ckd_categorical.duplicated(subset=['unique_id'])

# Count the total number of duplicate rows
num_duplicates1 = duplicates1.sum()

# Print the total number of duplicate rows
if num_duplicates1 > 0:
    print(f"Total number of Categorical duplicate rows: {num_duplicates1}")
    
    # Drop these duplicate rows
    ckd_categorical = ckd_categorical.drop_duplicates(subset=['unique_id'])
    print("Categorical Duplicate rows removed.")
else:
    print("No Categorical duplicate rows found.")

Total number of Numericsl duplicate rows: 30
Numerical Duplicate rows removed.
Total number of Categorical duplicate rows: 30
Categorical Duplicate rows removed.


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer:

Unique IDs Check: By using sets, you can easily find common and missing IDs.

Merge Operation: The how='outer' parameter ensures that all rows from both datasets are included, with NaN values where there is no match.

Result: The combined dataset will show all entries from both datasets. Rows with missing data in one of the datasets will have NaN values for those columns.

Are All Unique IDs Present?
If all unique IDs from one dataset are not present in the other, those rows will appear with NaN values in the columns from the dataset where the ID is missing.
This indicates that some records are only present in one of the datasets.

Example of Missing Rows:
If a row with unique_id = 123456 is only in df_categorical, it will appear in combined_df with NaN for columns from df_numerical. Similarly, if an ID is only in df_numerical, it will have NaN for columns from df_categorical.

In [109]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
unique_ids_numerical = set(ckd_numerical['unique_id'])
unique_ids_categorical = set(ckd_categorical['unique_id'])

# IDs present in both
common_ids = unique_ids_numerical.intersection(unique_ids_categorical)

# Merge datasets
combined_df = pd.merge(ckd_numerical, ckd_categorical, on='unique_id', how='outer')

# Print combined dataset
print(combined_df)
print(common_ids)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       103570  24.0  80.0  125.0    NaN   NaN  136.0  3.5  15.4  43.0  ...   
2       104675  65.0  80.0  215.0  133.0   2.5    NaN  NaN  13.2  41.0  ...   
3       105923  90.0  90.0  139.0   89.0   3.0  140.0  4.1  12.0  37.0  ...   
4       107297  60.0  70.0    NaN    NaN   NaN    NaN  NaN  16.4  43.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   
397     996412  70.0  60.0  120.0   40.0   0.5  140.0  4.6  16.0  43.0  ...   
398     997629  35.0   NaN   99.0   30.0   0.5  135.0  4.9  15.4  48.0  ...   
399     997876  64.0  70.0  107.0   15.0   NaN    NaN  NaN  12.8  38.0  ...   

           pc         pcc          ba  htn   dm  ca

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [110]:
# Calculate the percentage of rows that contain atleast one missing value
missing_values_count = combined_df.isnull().any(axis=1).sum()
total_rows = len(combined_df)
percentage_missing = (missing_values_count / total_rows) * 100

# Print %
print(f"Percentage of rows with at least one missing value: {percentage_missing:.2f}%")

# Drop these rows from the dataset
df_cleaned = combined_df.dropna()

# Print the Dataset
print(df_cleaned)

Percentage of rows with at least one missing value: 61.75%
     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
6       109053  64.0  60.0  106.0   27.0   0.7  150.0  3.3  14.4  42.0  ...   
7       114717  41.0  80.0  122.0   25.0   0.8  138.0  5.0  17.1  41.0  ...   
8       118191  42.0  70.0   93.0   32.0   0.9  143.0  4.7  16.6  43.0  ...   
9       118596  74.0  60.0   88.0   50.0   0.6  147.0  3.7  17.2  53.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
390     980291  62.0  70.0  122.0   42.0   1.7  136.0  4.7  12.6  39.0  ...   
391     991031  30.0  80.0   89.0   42.0   0.5  139.0  5.0  16.7  52.0  ...   
394     994377  71.0  70.0   81.0   18.0   0.8  145.0  5.0  14.7  44.0  ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  

## Part 3.4: Sort the dataset according to the Labels

In [111]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
sorted_df = df_cleaned.sort_values(by='Target').reset_index(drop=True)

# Print the dataset
print(sorted_df)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0  ...   
2       343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0  ...   
3       349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0  ...   
4       353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
148     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  44.0  ...   
149     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  42.0  ...   
150     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  50.0  ...   
151     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  50.0  ...   
152     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

           pc         pcc          ba  htn   dm  ca

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [112]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']

# Apply one-hot encoding to categorical columns
encoded_df = pd.get_dummies(sorted_df, columns=categorical_columns, dtype = 'int')

# Print the encoded dataset
print(encoded_df)

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0  ...   
2       343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0  ...   
3       349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0  ...   
4       353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
148     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  44.0  ...   
149     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  42.0  ...   
150     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  50.0  ...   
151     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  50.0  ...   
152     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer:

When encoding categorical data, using a single binary column for a variable with two categories (e.g., 'rbc') is equivalent to using two columns because one column can represent the presence or absence of a category. However, this approach doesn't work for variables with more than two categories.

A "type of condiment" feature with five categories requires five separate binary columns, one for each category. Using one-hot encoding for multi-category features ensures that the model does not infer any unintended relationships between categories, which could lead to biased predictions.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [113]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

for column in numerical_columns:
    mean = encoded_df[column].mean()
    std_dev = encoded_df[column].std()
    # Define the upper and lower bounds for outliers
    upper_bound = mean + (3 * std_dev)
    lower_bound = mean - (3 * std_dev)
    # Filter out the outliers
    encoded_df = encoded_df[(encoded_df[column] <= upper_bound) & (encoded_df[column] >= lower_bound)]

# Print the cleaned dataset
print(encoded_df)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
2       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
5       397388  63.0  100.0  280.0   35.0  3.2  143.0  3.5  13.0  40.0  ...   
7       438182  73.0   80.0  253.0  142.0  4.6  138.0  5.8  10.5  33.0  ...   
8       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
9       475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
148     486397  20.0   70.0  123.0   44.0  1.0  135.0  3.8  14.6  44.0  ...   
149     203466  22.0   60.0   97.0   18.0  1.2  138.0  4.3  13.5  42.0  ...   
150     482105  44.0   60.0   95.0   46.0  0.5  138.0  4.2  15.0  50.0  ...   
151     228849  58.0   80.0  100.0   50.0  1.2  140.0  3.5  14.0  50.0  ...   
152     996076  57.0   60.0  105.0   49.0  1.2  150.0  4.7  15.7  44.0  ...   

     cad_no  cad_yes  appet_good  appet_poor  pe_no

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [114]:
# Normalize the all Numerical Attributes in the dataset.
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

# Initialize the scaler
scaler = StandardScaler()

# Fit and transform the numerical columns
encoded_df[numerical_columns] = scaler.fit_transform(encoded_df[numerical_columns])

# Print the normalized dataset
print(encoded_df)



     unique_id       age        bp       bgr        bu        sc       sod  \
2       343710  0.810190  0.730553  1.146438  4.014305  1.690051 -0.959702   
5       397388  0.936367  2.776102  3.520558 -0.225995  1.218114  0.448743   
7       438182  1.567253  0.730553  2.921481  3.789156  2.161987 -0.431535   
8       474407 -1.713353  1.753327 -0.317973 -0.038371  0.206821 -2.720258   
9       475200  0.999456  2.776102  0.924557  0.486976  3.914895 -0.079424   
..         ...       ...       ...       ...       ...       ...       ...   
148     486397 -1.776441 -0.292221  0.037036  0.111728 -0.265116 -0.959702   
149     203466 -1.650264 -1.314996 -0.539853 -0.863916 -0.130277 -0.431535   
150     482105 -0.262316 -1.314996 -0.584229  0.186777 -0.602214 -0.431535   
151     228849  0.620924  0.730553 -0.473289  0.336877 -0.130277 -0.079424   
152     996076  0.557836 -1.314996 -0.362349  0.299352 -0.130277  1.681133   

          pot      hemo       pcv  ...  cad_no  cad_yes  appet_

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer:

Based on the provided dataset, there is one column that is not appropriate for modeling and predictions:
Column to Remove: 'unique_id'

Justification for exclusion:
The 'unique_id' column contains unique identifiers for each row. These identifiers don't carry any predictive information about chronic kidney disease.
Including 'unique_id' in the model could lead to overfitting, as the model might try to learn patterns based on these arbitrary identifiers.
Unique identifiers are typically used for data management and record keeping, not for predictive modeling.


In [115]:
#Remove that column
encoded_df = encoded_df.drop('unique_id', axis=1)  #(unique_id column)

# Reset the index
encoded_df = encoded_df.reset_index(drop=True)

# Print the dataset
print(encoded_df)

          age        bp       bgr        bu        sc       sod       pot  \
0    0.810190  0.730553  1.146438  4.014305  1.690051 -0.959702  1.373024   
1    0.936367  2.776102  3.520558 -0.225995  1.218114  0.448743 -1.303189   
2    1.567253  0.730553  2.921481  3.789156  2.161987 -0.431535  2.317570   
3   -1.713353  1.753327 -0.317973 -0.038371  0.206821 -2.720258 -1.303189   
4    0.999456  2.776102  0.924557  0.486976  3.914895 -0.079424  0.428478   
..        ...       ...       ...       ...       ...       ...       ...   
128 -1.776441 -0.292221  0.037036  0.111728 -0.265116 -0.959702 -0.830916   
129 -1.650264 -1.314996 -0.539853 -0.863916 -0.130277 -0.431535 -0.043795   
130 -0.262316 -1.314996 -0.584229  0.186777 -0.602214 -0.431535 -0.201219   
131  0.620924  0.730553 -0.473289  0.336877 -0.130277 -0.079424 -1.303189   
132  0.557836 -1.314996 -0.362349  0.299352 -0.130277  1.681133  0.585903   

         hemo       pcv      wbcc  ...  cad_no  cad_yes  appet_good  \
0   

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [116]:
# Export the dataframe to a new csv file
encoded_df.to_csv('cleaned_chronic_kidney_disease.csv', index=False)
print("CSV file exported successfully.")

# Export the dataframe to a new json file
encoded_df.to_json('cleaned_chronic_kidney_disease.json', orient='records')
print("JSON file exported successfully.")

CSV file exported successfully.
JSON file exported successfully.


# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

### Paste the markdown table here

|    |  age     |     bp     |     bgr     |     bu     |     sc     |     sod     |     pot     |     hemo    |     pcv     |     wbcc    | ... | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|----|----------|------------|-------------|------------|------------|-------------|-------------|-------------|-------------|-------------|-----|--------|---------|------------|------------|-------|--------|--------|---------|-------------|----------------|
| 0  | 0.810190 | 0.730553   | 1.146438    | 4.014305   | 1.690051   | -0.959702   | 1.373024    | -2.913221   | -2.905503   | 0.519658    | ... | 0      | 1       | 0          | 1          | 0     | 1      | 0      | 1       | 1           | 0              |
| 1  | 0.936367 | 2.776102   | 3.520558    | -0.225995  | 1.218114   | 0.448743    | -1.303189   | -0.607180   | -0.590673   | 0.785483    | ... | 0      | 1       | 1          | 0          | 1     | 0      | 1      | 0       | 1           | 0              |
| 2  | 1.567253 | 0.730553   | 2.921481    | 3.789156   | 2.161987   | -0.431535   | 2.317570    | -1.694935   | -1.603411   | -0.366426   | ... | 0      | 1       | 1          | 0          | 1     | 0      | 1      | 0       | 1           | 0              |
| 3  | -1.713353| 1.753327   | -0.317973   | -0.038371  | 0.206821   | -2.720258   | -1.303189   | -2.652160   | -3.050180   | 1.937392    | ... | 1      | 0       | 1          | 0          | 0     | 1      | 0      | 1       | 1           | 0              |
| 4  | 0.999456 | 2.776102   | 0.924557    | 0.486976   | 3.914895   | -0.079424   | 0.428478    | -2.826200   | -2.616150   | -0.233513   | ... | 1      | 0       | 1          | 0          | 0     | 1      | 1      | 0       | 1           | 0              |
| 5  | 0.747101 | -1.314996  | 3.698062    | -0.188470  | 0.206821   | -1.839980   | -2.090311   | -2.826200   | -2.760826   | 3.177909    | ... | 1      | 0       | 0          | 1          | 1     | 0      | 0      | 1       | 1           | 0              |
| 6  | 0.494747 | 1.753327   | 0.170164    | 2.475789   | 3.577798   | -1.663925   | 0.743327    | -2.304078   | -2.182119   | -0.720859   | ... | 1      | 0       | 1          | 0          | 1     | 0      | 1      | 0       | 1           | 0              |
| 7  | 1.314898 | -0.292221  | 2.056147    | 2.063016   | 3.308119   | -3.600537   | -0.673492   | -2.173547   | -2.326796   | 1.538654    | ... | 0      | 1       | 1          | 0          | 0     | 1      | 0      | 1       | 1           | 0              |
| 8  | -0.136138| -1.314996  | 0.924557    | 1.912917   | 1.285533   | 0.096632    | -0.516068   | -1.999507   | -2.326796   | 2.912084    | ... | 1      | 0       | 1          | 0          | 1     | 0      | 1      | 0       | 1           | 0              |
| 9  | 0.873279 | -0.292221  | 0.014848    | 0.036678   | 0.206821   | -0.783646   | 0.585903    | -0.781221   | -0.735350   | -0.056296   | ... | 1      | 0       | 1          | 0          | 1     | 0      | 1      | 0       | 1           | 0              |
| 10 | 0.431659 | 0.730553   | 2.056147    | 1.199946   | 1.690051   | -0.607591   | 0.900751    | -1.520894   | -1.458735   | -0.277817   | ... | 1      | 0       | 1          | 0          | 0     | 1      | 1      | 0       | 1           | 0              |
| 11 | 0.999456 | -1.314996  | 2.610848    | 0.637075   | 1.959729   | -0.607591   | 1.687873    | -2.130037   | -2.182119   | -0.233513   | ... | 1      | 0       | 0          | 1          | 0     | 1      | 1      | 0       | 1           | 0              |
| 12 | 1.441076 | -1.314996  | -0.073905   | 3.151235   | 2.633924   | -0.783646   | 0.900751    | -1.303343   | -1.314058   | 3.177909    | ... | 1      | 0       | 0          | 1          | 0     | 1      | 1      | 0       | 1           | 0              |
| 13 | 1.441076 | -0.292221  | 2.167088    | 1.537669   | 1.487792   | -1.311813   | 0.113630    | -1.738445   | -1.603411   | -1.075292   | ... | 0      | 1       | 1          | 0          | 1     | 0      | 1      | 0       | 1           | 0              |
| 14 | 0.684013 | 0.730553   | 4.030883    | -0.225995  | -0.062857  | -3.248425   | -1.303189   | -1.738445   | -1.314058   | 1.272829    | ... | 1      | 0       | 0          | 1          | 1     | 0      | 1      | 0       | 1           | 0              |





** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

Answer:

Verification and validation:
Always cross-check the output against the original data source.
Use statistical methods to compare distributions and summary statistics before and after conversion.
Implement automated data integrity checks.
Pay special attention to edge cases and unusual data points.

Use specialized tools:
Rely on purpose-built data processing libraries and tools (e.g., pandas, numpy) for data conversions rather than language models.
These tools are designed specifically for data manipulation and are less prone to errors.


Incremental processing:
Process data in smaller batches and verify each batch before proceeding.
This allows for easier error detection and correction.
Keep track of all transformations applied to the data.
This helps in tracing back any inconsistencies to their source.

Use multiple models or approaches:
Apply different methods or models to the same task and compare results.
Discrepancies can highlight potential hallucinations or errors.
Use try-except blocks to catch and log potential errors during conversion.

Documentation:
Clearly document all steps in the conversion process.
Include any assumptions made and limitations of the approach.

By implementing these strategies, you can significantly reduce the risk of data conversion errors and hallucinations when using language models or other AI tools for data processing tasks. Remember, the key is to always approach AI-assisted data conversion with a critical eye and robust verification processes.


## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [117]:
# Code to rename all the columns in the dataset

# Define the column mapping
column_mapping = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'class': 'class'
}

# Rename the columns
encoded_df = encoded_df.rename(columns=column_mapping)

# Display the first few rows to verify the changes
print(encoded_df)

          age  blood pressure  blood glucose random  blood urea  \
0    0.810190        0.730553              1.146438    4.014305   
1    0.936367        2.776102              3.520558   -0.225995   
2    1.567253        0.730553              2.921481    3.789156   
3   -1.713353        1.753327             -0.317973   -0.038371   
4    0.999456        2.776102              0.924557    0.486976   
..        ...             ...                   ...         ...   
128 -1.776441       -0.292221              0.037036    0.111728   
129 -1.650264       -1.314996             -0.539853   -0.863916   
130 -0.262316       -1.314996             -0.584229    0.186777   
131  0.620924        0.730553             -0.473289    0.336877   
132  0.557836       -1.314996             -0.362349    0.299352   

     serum creatinine    sodium  potassium  hemoglobin  packed cell volume  \
0            1.690051 -0.959702   1.373024   -2.913221           -2.905503   
1            1.218114  0.448743  -1.303

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [118]:
# Converted SQL to Pandas code
result = encoded_df.groupby(['Target_ckd', 'Target_notckd']).size().reset_index(name='count')

print(result)

   Target_ckd  Target_notckd  count
0           0              1    111
1           1              0     22


*Describe what the above code does here*

This pandas code accomplishes the same task as the SQL query. Here's what it does:

It groups the data by the 'Target' column, which in this dataset represents the class (either 'ckd' for chronic kidney disease or 'notckd' for not chronic kidney disease).

It counts the number of occurrences for each unique value in the 'Target' column.

The result is a new DataFrame with two columns: 'Target' and 'count'.

This query essentially provides a summary of how many instances there are for each class in the dataset. It allows us to see the distribution of chronic kidney disease cases versus non-chronic kidney disease cases in the dataset.